### ISSUE TITLE
naughtyHref bypass using 0xa0

### ISSUE THREAD

**Issue by potetisensei** [apostrophecms/sanitize-html#305]
Hello, I found the way of bypassing naughtyHref with "\xa0" in Safari.

I know there is a check for malicious protocols in naughtyHref, using [the OWASP page](https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet#Embedded_tab) as a reference.

However, today I realized the check is not sufficient.
Here is the PoC:
<img width="782" alt="スクリーンショット 2019-09-13 17 14 04" src="https://user-images.githubusercontent.com/1556747/64847405-e8e83700-d649-11e9-8639-f0b0071bc2d8.png">
<img width="936" alt="スクリーンショット 2019-09-13 15 55 58" src="https://user-images.githubusercontent.com/1556747/64847379-d79f2a80-d649-11e9-81fc-59403b91130a.png">
You can see that Safari pops an alert, and that sanitizeHtml fails in sanitizing this html.

So basically, I recommend you to prohibit protocols which contain non-ascii-printable characters(not just 0x00 to 0x20).
Actually I'm not familiar with web so much, so I can't see whether this proposal is appropriate(are there any specifications for protocols?)

p.s. This is a browser-specific problem.  Chrome and Firefox are not affected.

**Comment by boutell** [comment]
Well that stinks!

Thanks for letting me know about the issue.

Interested in doing a PR?

**Comment by dyang-branch** [comment]
Hi @boutell, we have some similar issue.  Is there a plan to push out a PR for it.

Thanks
Daniel

**Comment by potetisensei** [comment]
Sorry for late reply. If anyone would not like to handle this, then I should take it, but I'm busy now so I want you not to expect a fast fix especially if you mind the very obvious patch like allowing only English alphabets(I don't know why you mind it though).

**Comment by boutell** [comment]
Apps can register protocols on smartphones etc. Worried about future
compatibility too.

However rfc 2396 says:

Scheme names consist of a sequence of characters beginning with a lower
case letter and followed by any combination of lower case letters, digits,
plus ("+"), period ("."), or hyphen ("-").

The rfc also defines lower case letters as a-z only.

So I think it would be ok to write a simple fix that rejects any protocol
name that doesn't satisfy this simple policy.
